src/ActiveLogin.Authentication.BankId.Api/Models/SignRequest.cs

```diff
@@ -0,0 +1,139 @@
+        /// </param>
+        /// <param name="userNonVisibleData">
+        /// Data not displayed to the user. String. The value must be base 64-encoded. 1-200 000charactersafter base 64-encoding.
+        /// </param>
+        /// <param name="requirement">Requirements on how the auth or sign order must be performed.</param>
+        public SignRequest(string endUserIp, string personalIdentityNumber, string userVisibleData, string userNonVisibleData, Requirement requirement)
+        {
+            EndUserIp = endUserIp;
+            PersonalIdentityNumber = personalIdentityNumber;
+            Requirement = requirement;
+            UserVisibleData = EnsureBase64EncodedString(userVisibleData);
+            UserNonVisibleData = EnsureBase64EncodedString(userNonVisibleData);
+        }
+
+        /// <summary>
+        /// The user IP address as seen by RP. IPv4 and IPv6 is allowed.
+        /// Note the importance of using the correct IP address.It must be the IP address representing the user agent (the end user device) as seen by the RP.
+        /// If there is a proxy for inbound traffic, special considerations may need to be taken to get the correct address.
+        ///
+        /// In some use cases the IP address is not available, for instance for voice based services.
+        /// In this case, the internal representation of those systems IP address is ok to use.
+        /// </summary>
+        [DataMember(Name = "endUserIp")]
+        public string EndUserIp { get; private set; }
+
+        /// <summary>
+        /// The personal number of the user. 12 digits, century must be included (YYYYMMDDSSSC).
+        /// If the personal number is excluded, the client must be started with the AutoStartToken returned in the response.
+        /// </summary>
+        [DataMember(Name = "personalNumber", EmitDefaultValue = false)]
+        public string PersonalIdentityNumber { get; private set; }
+
+        /// <summary>
+        /// Requirements on how the auth or sign order must be performed.
+        /// </summary>
+        [DataMember(Name = "requirement", EmitDefaultValue = false)]
+        public Requirement Requirement { get; private set; }
+
+        /// <summary>
+        /// The text to be displayed and signed. String. The text can be formatted using CR, LF and CRLF for new lines.
+        /// The text must be encoded as UTF-8 and then base 64 encoded. 1--40 000 characters after base 64 encoding.
+        /// </summary>
+        [DataMember(Name = "userVisibleData")]
+        public string UserVisibleData { get; private set; }
+
+        /// <summary>
+        /// Data not displayed to the user. String. The value must be base 64-encoded. 1-200 000 characters after base 64-encoding.
+        /// </summary>
+        [DataMember(Name = "userNonVisibleData", EmitDefaultValue = false)]
+        public string UserNonVisibleData { get; private set; }
+
+        private static string EnsureBase64EncodedString(string value)
```

Comment:
I would like us to always expect the user to send in the data as a "normal string" and that the Base64 encoding is always done. That would be much more predictable in behavior. So, rename to `ToBase64String` and donät check if already Base64.

Reply:
If you want to base64 something other than a string you won' be able to though? In that case I'd suggest you change the input parameter type or add a overload with object userNonVisibleData at least which handles that?

Reply:
Very good point! Let me have a look on other implementations and get back.

The userVisibleData is always a string in the end (from what I can read), so that is more straight forward, but on the other hand it might be strange if we encode one but not the other.

Do you know what the userNonVisibleData is used for? Is it returened on collect in any way?

Reply:
I don't have a response at hand from the new endpoint, another project today, but this is from my notes when investigating the v4 soap endpoint. Unfortunately, in Swedish.
![bild](https://user-images.githubusercontent.com/17863113/52776886-e1749700-3042-11e9-8bbe-9c59eee920a1.png)

I hopefully have some time tonight to implement your suggestions and will give you a sample from the new endpoint. I'm also thinking about adding a Raw collect-method as well if you want to store the complete original response as is without deserializing.

Reply:
Maybe it would be easier just to change all input parameters to byte[] instead for the invisible part, Then the user can choose freely. It would simplify the overloads/constructors as well, not everything being string. How do you stand when it comes to optional parameters as well, it would simplify when it comes to the number of explicit constructors needed.

Reply:
Changed to byte[] now instead

Reply:
![bild](https://user-images.githubusercontent.com/17863113/52827043-7cad5100-30c3-11e9-9502-df04802c00fb.png)
Visible in bankidprompt: "test signatur" Non visible: new byte[1]

Fetched with CompletionData.GetSignatureXml()

Reply:
Thanks for the samples!